### PR TITLE
ErdKotlin 0.5.0: Implement ESDT

### PR DESCRIPTION
**Pull request by Alexandre-saddour**
This PR adds the following
- Implement ESDT API endpoints
  - Get all ESDT tokens for an address
  - Get balance for an address and an ESDT token
  - Get all issued ESDT tokens
  - Get ESDT token properties
  - Get special roles for a token
- Implement ESDT management (issue, mint, burn, etc..)
- Unit test all usecases related to ESDT
- update url in order to use the gateway.
- Clean code

### CODE REVIEW

File: src/main/java/com/elrond/erdkotlin/data/esdt/EsdtRepositoryImpl.kt [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+            )
+        )
+        return response.toEsdtProperties()
+    }
+
+    override fun getEsdtSpecialRoles(tokenIdentifier: String): EsdtSpecialRoles? {
+        val response = vmRepository.queryContract(
+            QueryContractInput(
+                scAddress = EsdtConstants.ESDT_SC_ADDR.bech32,
+                funcName = "getSpecialRoles",
```
Comment (ditzdragos) [inline src/main/java/com/elrond/erdkotlin/data/esdt/EsdtRepositoryImpl.kt]: Can we extract the function names in constants?

Comment (Alexandre-saddour) [reply src/main/java/com/elrond/erdkotlin/data/esdt/EsdtRepositoryImpl.kt]: Yes, will do :)

File: src/main/java/com/elrond/erdkotlin/domain/esdt/management/TransferEsdtUsecase.kt [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            if (funcArgs.isNotEmpty()) {
+                addAll(funcArgs.map { ScUtils.prepareArgument(it) })
+            }
+        }
+        return sendTransactionUsecase.execute(
+            Transaction(
+                sender = account.address,
+                receiver = receiver,
+                value = ESDT_TRANSACTION_VALUE,
+                gasLimit = 500000L, // TODO + <an appropriate amount for the method call>
```
Comment (ditzdragos) [inline src/main/java/com/elrond/erdkotlin/domain/esdt/management/TransferEsdtUsecase.kt:45]: I'll check with the team and provide the formula for the gasLimit

File: src/main/java/com/elrond/erdkotlin/domain/sc/ScUtils.kt [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+        if (!argUpCase.isDigitsOnly()){
+            throw IllegalArgumentException("unknown format for $arg")
+        }
+
+        return argUpCase.toBigInteger().toHex().toUpperCase(Locale.ROOT)
+    }
+
+    fun prepareBooleanArgument(key: String, value: Boolean) =
+        "${key.toHex()}@${value.toString().toHex()}"
```
Comment (ditzdragos) [inline src/main/java/com/elrond/erdkotlin/domain/sc/ScUtils.kt:27]: for HEX strings, you should make sure that they have an even length. If the length is an odd number, you need to add an `0`(zero)  as a prefix to make it even.

File: src/main/java/com/elrond/erdkotlin/data/esdt/EsdtRepositoryImpl.kt (hunk 1 above)
Comment (Alexandre-saddour) [reply src/main/java/com/elrond/erdkotlin/data/esdt/EsdtRepositoryImpl.kt]: Done ✅

**Review by ditzdragos — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Pull request merged**